### PR TITLE
Remove some redundant LANGUAGE pragmas and set other-extensions

### PR DESCRIPTION
**Pull request by hvr**
This way, cabal's solver detects that GHC 7.4.2 doesn't satisfy
other-extensions and informs us that a compiler is needed which
supports the LambdaCase extension.

### CODE REVIEW

File: cabal-install/Distribution/Client/ProjectPlanning.hs [hunk 1]
```diff
@@ -1,4 +1,4 @@
-{-# LANGUAGE CPP, RecordWildCards, NamedFieldPuns, RankNTypes #-}
+{-# LANGUAGE RecordWildCards, NamedFieldPuns, RankNTypes #-}
 {-# LANGUAGE ViewPatterns #-}
```
Comment (BardurArantsson) [inline cabal-install/Distribution/Client/ProjectPlanning.hs]: Given that e.g. ViewPatterns is now listed in other-extensions, shouldn't these also be removed? Or am I missing something?

**Comment by hvr** [comment]
@23Skidoo btw, this PR is against master; but before releasing cabal-install 2.0, you should check that the cabal file excludes older GHC versions which would fail to build it

File: cabal-install/Distribution/Client/ProjectPlanning.hs (hunk 1 above)
Comment (hvr) [inline cabal-install/Distribution/Client/ProjectPlanning.hs]: @23Skidoo that's btw a very common misconception... it's too late now, but I wonder if `depend-extensions` would have been a better name.

**Comment by 23Skidoo** [comment]
@hvr Via a minimum bound on `base`?

**Comment by hvr** [comment]
@23Skidoo yeah; that would be the simplest way to achieve this (in a pragmatic way - from a purist POV it's not principled of course :-) ).

**Comment by hvr** [comment]
This exposes a problem in the Travis job entry `GHCVER=7.4.2 SCRIPT=script CABAL_LIB_ONLY=YES TEST_OTHER_VERSIONS=YES`, as `cabal-install` isn't compatible w/ GHC 7.4.2:

```
The command "rm -rf dist-newstyle" exited with 0.
$ ./travis-${SCRIPT}.sh -j1
$ cabal update
Config file path source is default config file.
Config file /home/travis/.cabal/config not found.
Writing default configuration to /home/travis/.cabal/config
Downloading the latest package list from hackage.haskell.org
cabal update took 5 seconds.
whole job took 5 seconds so far.
----
/opt/happy/1.19.5/bin/happy
$ cabal unpack hackage-repo-tool-0.1.1
Unpacking to hackage-repo-tool-0.1.1/
cabal unpack hackage-repo-tool-0.1.1 took 0 seconds.
whole job took 5 seconds so far.
----
$ cabal new-build Cabal Cabal:unit-tests
Resolving dependencies...
cabal: Could not resolve dependencies:
next goal: cabal-install (user goal)
rejecting: cabal-install-2.1.0.0 (conflict: requires LambdaCase)
```

**Comment by ezyang** [comment]
Oh yeah, I guess previously we were saved by the fact that we only ever tried to build the library. But we always solved for cabal-install and so now it is failing.

Easy "fix" is to remove cabal-install directory prior to doing rest of CI.

**Comment by hvr** [comment]
@23Skidoo fyi, http://hackage.haskell.org/package/cabal-install-2.0.0.0/revisions/

**Comment by 23Skidoo** [comment]
@hvr Can you also apply that fix to the 2.0 branch?

**Comment by hvr** [comment]
@23Skidoo well, in order to apply this to the 2.0 we'd first have to resolve the test-suite issue that blocked this PR going in... otherwise the testsuite for 2.0 will just start failing :-)

File: cabal-install/Distribution/Client/ProjectPlanning.hs (hunk 1 above)
Comment (23Skidoo) [inline cabal-install/Distribution/Client/ProjectPlanning.hs]: IMO `depend-extensions` would be even more confusing.

**Comment by 23Skidoo** [comment]
> Btw, shall we also consider `cabal-version:>=1.18` for `cabal-install.cabal`?

OK, sure, though it'd nice to find out why the setup script segfaults.

**Comment by 23Skidoo** [comment]
> well, in order to apply this to the 2.0 we'd first have to resolve the test-suite issue that blocked this PR going in... 

Maybe we can have a separate `cabal.project` file for the library-only configuration?

**Comment by ezyang** [comment]
> OK, sure, though it'd nice to find out why the setup script segfaults.

Segfault?

**Comment by 23Skidoo** [comment]
@ezyang @hvr tried to run `cabal install cabal-install-2.0.0.0` with cabal-install 1.16 and it failed to build. Quoting the e-mail @hvr sent me:

>```
> $ PATH=/opt/ghc/7.6.3/bin:$PATH /opt/cabal/1.16/bin/cabal build
> Warning: cabal-install.cabal: Ignoring unknown section type: custom-setup
> [1 of 1] Compiling Main             ( Setup.hs, dist/setup/Main.o )
> Linking ./dist/setup/setup ...
> setup: Run the 'configure' command first.
> ```
>
> But if I change `build-type` to Simple, it works fine.
> 
> Looking deeper, it seems that `./dist/setup/setup configure` just
> crashes (exit-code 11 / SIGSEGV)... and thus no configure state is
> saved; but it's not clear to me why...

**Comment by gbaz** [comment]
for a code cleanup pass, i think this is way too bitrotted to rescue :-/

**Comment by Mikolaj** [comment]
I think we can return in a few ears, set Haskell2021 and use run stylish-haskell to remove all spurious pragmas.

**Pull request closed without merging**